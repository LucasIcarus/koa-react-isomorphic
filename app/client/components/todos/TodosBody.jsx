/* @flow */
/* eslint-disable react/no-array-index-key */
import React from "react";
import { List } from "immutable";
import partial from "lodash/partial";
import style from "./TodosBodyStyle.css";
import type {
  CompleteTodoActionType,
  RemoveTodoActionType,
  TodoType
} from "./types";

const TodosBody = ({
  todos,
  completeTodo,
  removeTodo
}: { // eslint-disable-line indent
  todos: List<TodoType>,
  completeTodo: CompleteTodoActionType,
  removeTodo: RemoveTodoActionType
}) => (
  <div className={`col-md-12 ${style.container}`}>
    <table className="table">
      <tbody>
        {todos.toJS().map((todo: TodoType, index: number) => {
          const text = todo.complete
            ? <s>{todo.text}</s>
            : <span>{todo.text}</span>;

          return (
            <tr key={index}>
              <td>
                <span>
                  {index + 1}
                </span>
              </td>
              <td>
                {text}
              </td>
              <td>
                <button
                  type="button"
                  className="btn btn-xs btn-success"
                  onClick={partial(completeTodo, index)}
                >
                  <i className="fa fa-check" />
                </button>
              </td>
              <td>
                <button
                  type="button"
                  className="btn btn-xs btn-danger"
                  onClick={partial(removeTodo, index)}
                >
                  <i className="fa fa-remove" />
                </button>
              </td>
            </tr>
          );
        })}
      </tbody>
    </table>
  </div>
);

export default TodosBody;
