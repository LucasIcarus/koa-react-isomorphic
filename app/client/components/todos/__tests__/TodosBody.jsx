import _ from "lodash";
import faker from "faker";
import td from "testdouble";
import React from "react";
import { mount } from "enzyme";
import TodosBody from "../TodosBody";

describe("Component: TodosBody", () => {
  const viewer = {
    todos: {
      edges: _(4)
        .range()
        .map(value => ({
          node: {
            id: value,
            text: `Todo ${value + 1}`,
            complete: false
          }
        }))
        .value()
    }
  };
  let Relay;
  let component;
  let randomUUID;

  beforeEach(() => {
    randomUUID = faker.random.uuid();
    component = mount(<TodosBody viewer={viewer} />);
    Relay = {
      Store: {
        commitUpdate: td.function()
      }
    };
    TodosBody.__Rewire__("Relay", Relay);
  });

  afterEach(() => {
    TodosBody.__ResetDependency__("Relay");
  });

  it("should display a list of todos", () => {
    const trComponents = component.find("tr");
    expect(trComponents).toHaveLength(viewer.todos.edges.length);
  });

  describe("# when click on delete button", () => {
    let removeTodoMutation;

    beforeEach(() => {
      removeTodoMutation = td.function();
      td
        .when(removeTodoMutation(td.matchers.anything()))
        .thenReturn({ randomUUID });
      TodosBody.__Rewire__("RemoveTodoMutation", removeTodoMutation);
    });

    afterEach(() => {
      TodosBody.__ResetDependency__("RemoveTodoMutation");
    });

    it('should call "removeTodo"', () => {
      const trComponents = component.find("tr");

      trComponents.forEach(tr => {
        const removeButton = tr.find(".btn-danger");
        removeButton.simulate("click");

        expect(removeButton).toBeDefined();
        td.verify(Relay.Store.commitUpdate({ randomUUID }));
      });
      expect(td.explain(Relay.Store.commitUpdate).callCount).toEqual(viewer.todos.edges.length);
    });
  });

  describe("# when click on complete button", () => {
    let completeTodoMutation;

    beforeEach(() => {
      completeTodoMutation = td.function();
      td
        .when(completeTodoMutation(td.matchers.anything()))
        .thenReturn({ randomUUID });
      TodosBody.__Rewire__("CompleteTodoMutation", completeTodoMutation);
    });

    afterEach(() => {
      TodosBody.__ResetDependency__("CompleteTodoMutation");
    });

    it('should call "completeTodo"', () => {
      const trComponents = component.find("tr");

      trComponents.forEach(tr => {
        const completeButton = tr.find(".btn-success");
        completeButton.simulate("click");

        expect(completeButton).toBeDefined();
        td.verify(Relay.Store.commitUpdate({ randomUUID }));
      });
      expect(td.explain(Relay.Store.commitUpdate).callCount).toEqual(viewer.todos.edges.length);
    });
  });
});
