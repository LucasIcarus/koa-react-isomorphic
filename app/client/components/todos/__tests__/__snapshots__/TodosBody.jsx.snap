// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TodosBody should display a list of todos 1`] = `
<TodosBody
  viewer={
    Object {
      "todos": Object {
        "edges": Array [
          Object {
            "node": Object {
              "complete": false,
              "id": 0,
              "text": "Todo 1",
            },
          },
          Object {
            "node": Object {
              "complete": false,
              "id": 1,
              "text": "Todo 2",
            },
          },
          Object {
            "node": Object {
              "complete": false,
              "id": 2,
              "text": "Todo 3",
            },
          },
          Object {
            "node": Object {
              "complete": false,
              "id": 3,
              "text": "Todo 4",
            },
          },
        ],
      },
    }
  }
>
  <div
    className="col-md-12 container"
  >
    <table
      className="table"
    >
      <tbody>
        <tr>
          <td>
            <span>
              1
            </span>
          </td>
          <td>
            <span>
              Todo 1
            </span>
          </td>
          <td>
            <button
              className="btn btn-xs btn-success"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-check"
              />
            </button>
          </td>
          <td>
            <button
              className="btn btn-xs btn-danger"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-remove"
              />
            </button>
          </td>
        </tr>
        <tr>
          <td>
            <span>
              2
            </span>
          </td>
          <td>
            <span>
              Todo 2
            </span>
          </td>
          <td>
            <button
              className="btn btn-xs btn-success"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-check"
              />
            </button>
          </td>
          <td>
            <button
              className="btn btn-xs btn-danger"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-remove"
              />
            </button>
          </td>
        </tr>
        <tr>
          <td>
            <span>
              3
            </span>
          </td>
          <td>
            <span>
              Todo 3
            </span>
          </td>
          <td>
            <button
              className="btn btn-xs btn-success"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-check"
              />
            </button>
          </td>
          <td>
            <button
              className="btn btn-xs btn-danger"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-remove"
              />
            </button>
          </td>
        </tr>
        <tr>
          <td>
            <span>
              4
            </span>
          </td>
          <td>
            <span>
              Todo 4
            </span>
          </td>
          <td>
            <button
              className="btn btn-xs btn-success"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-check"
              />
            </button>
          </td>
          <td>
            <button
              className="btn btn-xs btn-danger"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-remove"
              />
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</TodosBody>
`;

exports[`Component: TodosBody should render component 1`] = `
<TodosBody
  viewer={
    Object {
      "todos": Object {
        "edges": Array [
          Object {
            "node": Object {
              "complete": false,
              "id": 0,
              "text": "Todo 1",
            },
          },
          Object {
            "node": Object {
              "complete": false,
              "id": 1,
              "text": "Todo 2",
            },
          },
          Object {
            "node": Object {
              "complete": false,
              "id": 2,
              "text": "Todo 3",
            },
          },
          Object {
            "node": Object {
              "complete": false,
              "id": 3,
              "text": "Todo 4",
            },
          },
        ],
      },
    }
  }
>
  <div
    className="col-md-12 container"
  >
    <table
      className="table"
    >
      <tbody>
        <tr>
          <td>
            <span>
              1
            </span>
          </td>
          <td>
            <span>
              Todo 1
            </span>
          </td>
          <td>
            <button
              className="btn btn-xs btn-success"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-check"
              />
            </button>
          </td>
          <td>
            <button
              className="btn btn-xs btn-danger"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-remove"
              />
            </button>
          </td>
        </tr>
        <tr>
          <td>
            <span>
              2
            </span>
          </td>
          <td>
            <span>
              Todo 2
            </span>
          </td>
          <td>
            <button
              className="btn btn-xs btn-success"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-check"
              />
            </button>
          </td>
          <td>
            <button
              className="btn btn-xs btn-danger"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-remove"
              />
            </button>
          </td>
        </tr>
        <tr>
          <td>
            <span>
              3
            </span>
          </td>
          <td>
            <span>
              Todo 3
            </span>
          </td>
          <td>
            <button
              className="btn btn-xs btn-success"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-check"
              />
            </button>
          </td>
          <td>
            <button
              className="btn btn-xs btn-danger"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-remove"
              />
            </button>
          </td>
        </tr>
        <tr>
          <td>
            <span>
              4
            </span>
          </td>
          <td>
            <span>
              Todo 4
            </span>
          </td>
          <td>
            <button
              className="btn btn-xs btn-success"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-check"
              />
            </button>
          </td>
          <td>
            <button
              className="btn btn-xs btn-danger"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-remove"
              />
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</TodosBody>
`;
